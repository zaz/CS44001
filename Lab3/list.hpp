//
// node class template used to contruct lists
//
// Mikhail Nesterenko
// Modified by: Zaz Brown
//

#ifndef LIST_HPP_
#define LIST_HPP_

#include <memory>

// the pointer part of every list is the same
// lists differ by data part
template <typename T>
class node{
public:
  node(): next_(nullptr) {}

  ~node() {
    // do not destroy next_, since this will delete all further nodes
  }

  // functions can be inlined
  T getData() const {return data_;}
  void setData(const T& data){data_=data;}

  // or can be defined outside
  std::shared_ptr<node<T>> getNext() const;
  void setNext(std::shared_ptr<node<T>> const);

private:
  T data_;
  std::shared_ptr<node<T>> next_;
};

//
// member functions for node
//
template <typename T>
std::shared_ptr<node<T>> node<T>::getNext() const {
  return next_;
}

template<typename T>
void node<T>::setNext(std::shared_ptr<node<T>> const next){
   next_ = next;
}

#endif // LIST_HPP_
