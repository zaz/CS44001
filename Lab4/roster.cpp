//
// Student roster
//
// Mikhail Nesterenko
// Modified by Zaz Brown
//

#include <fstream>
#include <iostream>
#include <vector>
#include <list>
#include <string>
#include <cstdlib>
#include <algorithm>
#include <filesystem>  // for proper handling of file paths

using std::ifstream;
using std::string; using std::getline;
using std::list; using std::vector;
using std::cout; using std::endl; using std::cerr;
using std::move;

// read a file of names into a list of strings
void readRoster(list<string>& roster, string fileName) {
   ifstream course(fileName);
   string first, last;
   // For each line in the file, read the first and second whitespace-separated
   // words into first and last, respectively.
   while(course >> first >> last)
      roster.push_back(first + ' ' + last);
   course.close();
}

// print a list
void printRoster(const list<string>& roster, bool stderr = false) {
   if (stderr) for(const auto& str : roster) cerr << str << '\n';
   else for(const auto& str : roster) cout << str << '\n';
}

// print student entries
void printEntries(const list<list<string>>& entries) {
   for (const auto& student : entries) {
      for (const auto& str : student)
         cout << str;
      cout << '\n';
      }
}

// XXX UGLY. We should use maps instead.
// create studentEntries, a list of lists of: "name: [courses enrolled...]"
list<list<string>> getStudentEntries(const vector<list<string>>& courseStudents,
                                     const list<string>& allStudents,
                                     char* argv[]) {
   list<list<string>> studentEntries;
   for (auto& student : allStudents) {
      list<string> entry;
      entry.push_back(student + ":");
      for (auto& lst : courseStudents) {
         if (std::find(lst.begin(), lst.end(), student) != lst.end()) {
            std::filesystem::path csI = argv[&lst - &courseStudents[0] + 1];
            // get only the filename, not the whole path, and not the extension
            csI = csI.filename();
            csI.replace_extension();
            entry.push_back(" " + string(csI));
         }
      }
      studentEntries.push_back(move(entry));
   }
   return studentEntries;
}

int main(int argc, char* argv[]) {
   if (argc <= 1) {
      cerr << "usage: " << argv[0]
      << " list of courses, dropouts last"
      << endl;
      exit(1);
   }

   // vector of courses of students
   vector<list<string>> courseStudents;
   for (int i=1; i < argc-1; ++i) {
      list<string> roster;
      readRoster(roster, argv[i]);
      // roster now contains a list of students in course argv[i]
      cerr << "\n\n" << argv[i] << "\n";
      printRoster(roster, true);
      courseStudents.push_back(move(roster));
   }

   // reading in dropouts
   list<string> dropouts;
   readRoster(dropouts, argv[argc-1]);
   cerr << "\n\n dropouts \n";
   printRoster(dropouts, true);

   // master list of students
   list<string> allStudents;
   for (const auto& lst : courseStudents)
      // copy list instead of splicing:
      allStudents.insert(allStudents.end(), lst.begin(), lst.end());

   auto studentEntries = getStudentEntries(courseStudents, allStudents, argv);

   cout << "All students\nfirst name last name: courses enrolled";
   printEntries(studentEntries);

   // sorting master list
   allStudents.sort();
   cerr << "\n\n all students sorted \n";
   printRoster(allStudents, true);

   // eliminating duplicates
   allStudents.unique();
   cerr << "\n\n all students, duplicates removed \n";
   printRoster(allStudents, true);

   // removing individual dropouts
   for (const auto& str : dropouts)
      allStudents.remove(str);

   cout << "All students, dropouts removed and sorted\n";
   cout << "first name last name:  courses enrolled\n";
   printRoster(allStudents);
}
