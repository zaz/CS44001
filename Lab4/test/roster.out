All students, dropouts removed and sorted
first name last name:  courses enrolled
Andrew Turner: cs3
Betty Torres: cs1
Brian Sanchez: cs1 cs4
Catherine Morgan: cs2
Charles Martin: cs1
Craig Jackson: cs4
Cynthia Jenkins: cs1 cs3
David Collins: cs1
Doris Howard: cs1
Elizabeth Green: cs3 cs4
Elizabeth Jones: cs1
Gerald Edwards: cs1 cs2 cs3 cs4
John Jenkins: cs2
Jonathan Baker: cs1
Joyce Rodriguez: cs2
Juan Cox: cs1 cs4
Judy Miller: cs4
Julie Perry: cs2
Kathleen Anderson: cs1 cs3
Linda Johnson: cs1 cs2
Lori Walker: cs2 cs3
Louis Gray: cs4
Louis Harris: cs3
Maria Kelly: cs1 cs4
Marilyn Phillips: cs2
Mark Russell: cs1 cs2 cs4
Martin Taylor: cs1
Mary Price: cs1
Nicholas Peterson: cs4
Peter Hall: cs2 cs3 cs4
Peter Young: cs2 cs3
Rachel Peterson: cs2
Rachel Rivera: cs1
Stephanie Watson: cs1
Tina Ward: cs3
